src/coprocessor/coprocessor.rs

```diff
@@ -0,0 +1,1142 @@
+// Copyright 2018 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::time::Duration;
+
+use futures::sync::mpsc;
+use futures::{future, stream, Future, Stream};
+use protobuf::{CodedInputStream, Message};
+
+use kvproto::{coprocessor as coppb, errorpb, kvrpcpb};
+use tipb::analyze::{AnalyzeReq, AnalyzeType};
+use tipb::checksum::{ChecksumRequest, ChecksumScanOn};
+use tipb::executor::ExecType;
+use tipb::select::DAGRequest;
+
+use server::readpool::{self, ReadPool};
+use server::Config;
+use storage::{self, Engine};
+use util::Either;
+
+use coprocessor::dag::executor::ExecutorMetrics;
+use coprocessor::metrics::*;
+use coprocessor::tracker::Tracker;
+use coprocessor::util as cop_util;
+use coprocessor::*;
+
+const OUTDATED_ERROR_MSG: &str = "request outdated.";
+const BUSY_ERROR_MSG: &str = "server is busy (coprocessor full).";
+
+pub struct Coprocessor<E: Engine> {
```

Comment:
I remember that @BusyJay disagreed with the Coprocessor Name here. 

seem we should use a better name.